source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1255,10 +1256,40 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::connPool(
     option->hashKey(hash_key);
   }
 
-  bool have_transport_socket_options = false;
-  if (context && context->upstreamTransportSocketOptions()) {
-    context->upstreamTransportSocketOptions()->hashKey(hash_key);
-    have_transport_socket_options = true;
+  Network::TransportSocketOptionsSharedPtr transport_socket_options =
+      context ? context->upstreamTransportSocketOptions() : nullptr;
+
+  // If configured to do so, we override the ALPN to use for the upstream connection to match the
+  // selected protocol.
+  if (cluster_info_->features() &
+      ClusterInfo::Features::PIN_UPSTREAM_ALPN_TO_SELECTED_HTTP_PROTOCOL) {
+    std::vector<std::string> alpn;
+    switch (protocol) {
```

Comment:
there might also be a desire to make this more configurable to allow for custom protocol, e.g. istio uses stuff like `istio-h2` to denote "HTTP/2 protocol to an istio sidecar" - so maybe this should be configurable per protocol instead of hard coded?